drivers/pwm/pwm_rpi_pico.c

```diff
@@ -0,0 +1,178 @@
+	return (float)cfg->slice_configs[slice].integral +
+		(float)cfg->slice_configs[slice].integral / 16.0;
+}
+
+static inline int pwm_rpi_channel_to_slice(uint32_t channel)
+{
+	return channel / 2;
+}
+
+static inline int pwm_rpi_channel_to_pico_channel(uint32_t channel)
+{
+	return channel % 2;
+}
+
+static int pwm_rpi_get_cycles_per_sec(const struct device *dev, uint32_t ch, uint64_t *cycles)
+{
+	float f_clock_in;
+	int slice = pwm_rpi_channel_to_slice(ch);
+
+	if (ch >= PWM_RPI_NUM_CHANNELS) {
+		return -EINVAL;
+	}
+
+	f_clock_in = (float)sys_clock_hw_cycles_per_sec();
+
+	/* No need to check for divide by 0 since the minimum value of
+	 * pwm_rpi_get_clkdiv is 1
+	 */
+	*cycles = (uint64_t)(f_clock_in / pwm_rpi_get_clkdiv(dev, slice));
+	return 0;
+}
+
+/* The pico_sdk only allows setting the polarity of both channels at once.
+ * This is a convenience function to make setting the polarity of a single
+ * channel easier.
+ */
+static void pwm_rpi_set_channel_polarity(const struct device *dev, int slice,
+					 int pico_channel, bool inverted)
+{
+	const struct pwm_rpi_config *cfg = dev->config;
+
+	bool pwm_polarity_a = (cfg->pwm_controller->slice[slice].csr & PWM_CH0_CSR_A_INV_BITS) > 0;
+	bool pwm_polarity_b = (cfg->pwm_controller->slice[slice].csr & PWM_CH0_CSR_B_INV_BITS) > 0;
+
+	if (pico_channel == PWM_CHAN_A) {
+		pwm_polarity_a = inverted;
+	} else if (pico_channel == PWM_CHAN_B) {
+		pwm_polarity_b = inverted;
+	}
+
+	pwm_set_output_polarity(slice, pwm_polarity_a, pwm_polarity_b);
+}
+
+static int pwm_rpi_pin_set(const struct device *dev, uint32_t ch, uint32_t period_cycles,
+			   uint32_t pulse_cycles, pwm_flags_t flags)
+{
+	if (ch >= PWM_RPI_NUM_CHANNELS) {
+		return -EINVAL;
+	}
+
+	if (period_cycles > PWM_RPI_PICO_COUNTER_TOP_MAX ||
+	    pulse_cycles > PWM_RPI_PICO_COUNTER_TOP_MAX) {
+		return -EINVAL;
+	}
+
+	int slice = pwm_rpi_channel_to_slice(ch);
+
+	/* this is the channel within a pwm slice */
+	int pico_channel = pwm_rpi_channel_to_pico_channel(ch);
+
+	pwm_rpi_set_channel_polarity(dev, slice, pico_channel,
+				     (flags & PWM_POLARITY_MASK) == PWM_POLARITY_INVERTED);
+	pwm_set_wrap(slice, period_cycles);
+	pwm_set_chan_level(slice, ch, pulse_cycles);
+
+	return 0;
+};
+
+struct pwm_driver_api pwm_rpi_driver_api = {
+	.get_cycles_per_sec = pwm_rpi_get_cycles_per_sec,
+	.pin_set = pwm_rpi_pin_set,
+};
+
+static int pwm_rpi_init(const struct device *dev)
+{
+	const struct pwm_rpi_config *cfg = dev->config;
+	pwm_config slice_cfg;
+	int err;
+
+	err = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err) {
+		LOG_ERR("Failed to configure pins for PWM. err=%d", err);
+		return err;
+	}
+
+	for (int i = 0; i < NUM_PWM_SLICES; i++) {
+		slice_cfg = pwm_get_default_config();
+		pwm_config_set_clkdiv_mode(&slice_cfg, PWM_DIV_FREE_RUNNING);
+
+		pwm_init(i, &slice_cfg, false);
+
+		pwm_set_clkdiv_int_frac(i,
+					cfg->slice_configs[i].integral,
+					cfg->slice_configs[i].frac);
+		pwm_set_enabled(i, true);
```

Comment:
a general comment, not related to this PR: I find RPi HAL really confusing since it uses the same Zephyr prefix (happens in other areas as well), I'd personally use register access, I think it would lead to more readable drivers + less overhead.